modules/module_service.go

```diff
@@ -97,14 +104,19 @@ func (w *moduleService) run(serviceContext context.Context) error {
 
 func (w *moduleService) stop(_ error) error {
 	var err error
-	if w.service.State() == services.Running {
+	switch w.service.State() {
+	case services.Running:
 		// Only wait for other modules, if underlying service is still running.
 		w.waitForModulesToStop()
-
+		// Then proceed to also stop this service if we didn't wait it out during moduleService.start.
+		fallthrough
+	case services.Starting:
+		// If upstream services can tolerate this service being failed and/or not started, and they have already started,
+		// they should track the state of this service and change state accordingly.
 		level.Debug(w.logger).Log("msg", "stopping", "module", w.name)
 
 		err = services.StopAndAwaitTerminated(context.Background(), w.service)
```

Comment:
Does this mean we could end up calling the stop function while the start is still running? I think that would be violation of the contract.

Reply:
From Dimitar on Slack:

> regarding your comment: since we’re using the services.BasicService abstraction to start and stop it will manage the state of the underlying w.service. When we call StopAndAwaitTerminated, this calls w.service.StopAsync() . Then if w.service is still running its startingFn, the context to that will be cancelled. And depending on the return error of that startingFn the BasicService will either run stoppingFn or will exit directly